lib/loader/index.js

```diff
@@ -0,0 +1,254 @@
+ *
+ * This module is derived from yepnope.js 2.0.0:
+ * Copyright (c) 2014, Alex Sexton
+ * Copyright (c) 2014, Ralph Holzmann
+ * All rights reserved.
+ * New BSD license:
+ * https://github.com/SlexAxton/yepnope.js/blob/master/LICENSE.md
+ *
+ * @module
+ */
+
+var global = require('../global');
+var doc = global.document;
+
+var DEFAULT_TIMEOUT = 10000;
+
+function getFirstScript () {
+  return doc.getElementsByTagName('script')[0];
+}
+
+function isFileReady (readyState) {
+  // the file is ready if there is no readyState, or if it has reached a
+  // terminal state.
+  return !readyState ||
+    readyState === 'loaded' ||
+    readyState === 'complete' ||
+    readyState === 'uninitialized';
+}
+
+function validateArguments (url, options, callback) {
+  if (!url || typeof url !== 'string') {
+    throw new Error('`url` must be a string');
+  }
+
+  if (typeof options.timeout !== 'number') {
+    throw new Error('`options.timeout` must be a number');
+  }
+
+  if (callback && typeof callback !== 'function') {
+    throw new Error('`callback` must be a function');
+  }
+}
+
+module.exports = {
+  /**
+   * Load a script from a URL
+   *
+   * The script is loaded on a later turn of the event loop. If a callback is
+   * provided, it will be invoked on success in all supported browsers. The
+   * callback will execute after the script.
+   *
+   * In modern IE and other browsers, the callback will be invoked with an error
+   * on failure as well. A timeout in milliseconds may be provided as
+   * `options.timeout`. If this timeout expires before loading fails or
+   * succeeds, the callback will be invoked with an error.
+   *
+   * @param {string} url - URL of the script
+   * @param {object} [options]
+   * @param {object} options.attributes - an object containing key/value pairs
+   *  to be used as attributes on the created `script` element (optional)
+   * @param {number} options.timeout - timeout in milliseconds (defaults to
+   *  10000ms)
+   * @param {function} [callback] - node-style callback
+   */
+  loadScript: function (url, options, callback) {
+    if (typeof options === 'function') {
+      callback = options;
+      options = null;
+    }
+
+    options = options || {};
+    options.timeout = options.timeout || DEFAULT_TIMEOUT;
+
+    validateArguments(url, options, callback);
+
+    var script = doc.createElement('script');
+    var done = false;
+    var attr;
+
+    if (options.attributes) {
+      for (attr in options.attributes) {
+        script.setAttribute(attr, options.attributes[attr]);
+      }
+    }
+
+    function cleanUp (err) {
+      done = true;
+      clearTimeout(timeoutHandle);
+      script.onload = script.onreadystatechange = script.onerror = null;
+      script.parentNode.removeChild(script);
+
+      if (callback) {
+        try {
+          callback(err);
+        }
+        catch (e) {
+          // FIXME: report this error once analytics support is available
```

Comment:
Maybe not applicable to this PR specifically, but surely there's something we can do to avoid swallowing these exceptions completely? I'm happy to open a PR if someone has an idea about how to do that (logging, wrapping the error, etc.).